### PR TITLE
Fix copying ntfs dacl and sacl when they are nil

### LINKED ISSUE #1992: Template resource fails updating some existing files
### Version:
11.14.6

### Environment:
Windows 7 x32

### Scenario:
Use the template resource to updated a pre-created file in Program Files. 

### Steps to Reproduce:
I have not figured out a way to create a file that causes this issue. I have just found some existing files that cause this issue.

Here is the code I am using that causes the issue:

    template "C:\\Program Files\\test\\example.ini" do
        source "example.ini.erb"
        variables(
            :test => 'foo',
            :things => ['foo', 'bar']
        )
    end

This is the template:
    
    this is test=<%= @test %>

    Things go here
    <% @things.each_with_index do |data, index| %>
    <%= "#{index}=#{data}" %>
    <% end %>
    
### Expected Result:
I am expecting the file to be updated with the output from the template file. 

### Actual Result:
When `C:\\Program Files\\test\\example.ini` already exists (with whatever properties cause this), the chef client throws `NoMethodError: private method 'select' called for nil:NilClass`. If the file is missing, chef can create it just fine. If the file is created manually on the machine, chef can update it without issues. I believe the issue comes in since I am copying the file from another machine so it must lack some attributes/properties or something. 

    3:09-04:00] DEBUG: Re-raising exception: NoMethodError - template[C:\Program Files\test\example.ini] (test::default line 10) had an error: NoMethodError: private method `select' called for nil:NilClass
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/file_content_management/deploy/mv_windows.rb:71:in `deploy'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/provider/file.rb:341:in `update_file_contents'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/provider/file.rb:372:in `block in do_contents_changes'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/mixin/why_run.rb:52:in `call'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/mixin/why_run.rb:52:in `add_action'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/provider.rb:156:in `converge_by'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/provider/file.rb:371:in `do_contents_changes'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/provider/file.rb:121:in `action_create'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/provider.rb:121:in `run_action'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource.rb:648:in `run_action'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/runner.rb:49:in `run_action'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/runner.rb:81:in `block (2 levels) in converge'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/runner.rb:81:in `each'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/runner.rb:81:in `block in converge'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource_collection.rb:98:in `block in execute_each_resource'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/resource_collection.rb:96:in `execute_each_resource'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/runner.rb:80:in `converge'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/client.rb:345:in `converge'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/client.rb:431:in `do_run'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/client.rb:227:in `run'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/application.rb:236:in `run_chef_client'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/application/client.rb:338:in `block in run_application'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/application/client.rb:327:in `loop'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/application/client.rb:327:in `run_application'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/lib/chef/application.rb:55:in `run'
    c:/opscode/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.14.6-x86-mingw32/bin/chef-client:26:in `<top (required)>'
    c:/opscode/chef/bin/chef-client:23:in `load'
    c:/opscode/chef/bin/chef-client:23:in `<main>'

### PR DESCRIPTION
**Pull request by SteveLowe**
Fixes #1992

### CODE REVIEW

File: lib/chef/file_content_management/deploy/mv_windows.rb [hunk 1]
```diff
@@ -63,12 +63,16 @@ def deploy(src, dst)
             raise Chef::Exceptions::WindowsNotAdmin, "can not get the security information for '#{dst}' due to missing Administrator privileges."
           end
 
-          if dst_sd.dacl_present?
-            apply_dacl = ACL.create(dst_sd.dacl.select { |ace| !ace.inherited? })
+          dacl_present = false
```
Comment (lamont-granquist) [inline lib/chef/file_content_management/deploy/mv_windows.rb]: you might want to just do this which is a little more straightforwards:

```
dacl_present = dst_sd.dacl_present? && !dst_sd.dacl.nil?
if dacl_present
   apply_dacl = ACL.create(dst_sd.dacl.select { |ace| !ace.inherited? })
end
```

**Comment by lamont-granquist** [comment]
This needs a spec test, although I'm not sure what we've got there for this, we may be relying on file provider unit and functional tests to test this code.

**Comment by jdmundrawala** [comment]
I'm not sure about the correctness of this. I believe it is valid to have a NULL DACL, and that is different that an empty DACL. From https://msdn.microsoft.com/en-us/library/windows/desktop/aa379286%28v=vs.85%29.aspx
```
If the discretionary access control list (DACL) that belongs to an object's security descriptor is set to NULL, a null DACL is created. A null DACL grants full access to any user that requests it; normal security checking is not performed with respect to the object. A null DACL should not be confused with an empty DACL. An empty DACL is a properly allocated and initialized DACL that contains no access control entries (ACEs). An empty DACL grants no access to the object it is assigned to.
```
It's not clear what that means for SACL.

cc @chef/client-windows

**Comment by lamont-granquist** [comment]
some of the commits here look odd, you might want to rebase on top of master and force push your branch.

**Comment by SteveLowe** [comment]
@lamont-granquist  - thanks, I thought I may have done something wrong there.

@jdmundrawala - good point. I guess this is going to need to be a bigger change than I thought.
I'll find some time soon to read up on NTFS and find a better solution.

**Comment by jdmundrawala** [comment]
:+1: Verified that it solves the problem

**Comment by jdmundrawala** [comment]
Rebased at https://github.com/chef/chef/pull/3471

**Comment by btm** [comment]
Closing in favor of the rebase in #3471.

**Pull request closed without merging**